### ISSUE TITLE
`from_dlpack` ignores stride information

### ISSUE THREAD

**Issue by pabloferz** [google/jax#4196]
Trying to pass a DLPack tensor to JAX I noticed that the strides information gets lost. Is this a bug or a limitation of the library?

E.g., if I have a `dltensor` that corresponds to a slice from say CuPy and I export it to DLPack and import it to JAX, I see the following:

```python
import cupy
import jax

from jax.dlpack import from_dlpack
from jax.config import config

config.update('jax_enable_x64', True)

a = cupy.arange(6).reshape((2, 3))
# array([[0, 1, 2],
#        [3, 4, 5]])
b = a[:, 0:2]
# array([[0, 1],
#        [3, 4]])
c = from_dlpack(b.toDlpack())
# DeviceArray([[0, 1],
#              [2, 3]], dtype=int64)
```

Where I was expecting:
```python
# DeviceArray([[0, 1],
#              [3, 4]], dtype=int64)
```

**Comment by jakevdp** [comment]
Thanks for the report - I can reproduce this. You're probably correct that it's a bug related to ignoring strides.

**Comment by jakevdp** [comment]
It looks like this is a problem in XLA rather than in JAX: the buffer produced by `xla_client` is incorrect:
```python
from jax.lib import xla_bridge, xla_client
import cupy

a = cupy.arange(6).reshape((2, 3))
b = a[:, :2].toDlpack()

backend = xla_bridge.get_backend()
client = getattr(backend, "client", backend)

buf = xla_client._xla.dlpack_managed_tensor_to_buffer(b, client)
print(buf.to_py())
# array([[0, 1],
#        [2, 3]], dtype=int64)
```

**Comment by jakevdp** [comment]
Trying to rule-out that the bug may be coming from cupy; going from cupy to torch looks OK:
```python
import cupy
import torch.utils.dlpack
a = cupy.arange(6).reshape((2, 3))
b = a[:, :2].toDlpack()
torch.utils.dlpack.from_dlpack(b)
# tensor([[0, 1],
#         [3, 4]], device='cuda:0')
```

**Comment by zhangqiaorjc** [comment]
This is currently a limitation of the library. 

JAX tries to convert the DLPack tensor (strided representation) to a XLA buffer ([minor_to_major representation](https://www.tensorflow.org/xla/shapes#minor-to-major_dimension_ordering)).

The minor_to_major representation can't represent a slice of a tensor. In this case, the shape after slicing is (2,2) and minor_to_major is (1,0), and the underlying data buffer is [0,1,2,3,4,5], which represents [[0,1],[2,3]], a wrong slice of the original tensor a.

_xla.dlpack_managed_tensor_to_buffer is supposed to [catch this error and return a RuntimeException ](https://github.com/tensorflow/tensorflow/blob/32cf223c2b8b578e289cd9f27041e4738b5710ed/tensorflow/compiler/xla/python/dlpack.cc#L182), but there's a bug that prevented it. The bug is [fixed](https://github.com/tensorflow/tensorflow/commit/28a2d0a1a83b3dfa3c903ce645aa9d319a4fd5ef) and a new release of JAX will return a RuntimeException.

There's a simple workaround by materializing the tensor after slicing with a deep copy before calling toDlpack(). Could you try that?

**Comment by pabloferz** [comment]
@jakevdp and @zhangqiaorjc thanks for looking into this.

A deep copy works, but I was trying to avoid copying the data and work with the underlying memory. Reading over 'minor_to_major representation' link makes me wonder whether a slice can be represented as a Layout with padding (I have no idea how padding is treated in general by XLA, probably it is there only for certain operations, maybe FFTs or what not).

**Comment by zhangqiaorjc** [comment]
@hawkinsp suggested "passing the unsliced array they have to jax and then use a jax operation to slice it". He may know more details

**Comment by mattjj** [comment]
I think Peter's idea is something like, instead of doing this:

```python
from jax.dlpack import from_dlpack
import cupy

a = cupy.arange(6).reshape((2, 3))
b = a[:, :2].toDlpack()  # this is slicing *first*, before passing to jax

c = from_dlpack(b)
```

to do this:

```python
from jax.dlpack import from_dlpack
import cupy

a = cupy.arange(6).reshape((2, 3))
b = a.toDlpack()  # no slice

c = from_dlpack(b)[:, :2]  # slice the JAX DeviceArray instead
```

**Comment by pabloferz** [comment]
That also works but still creates a copy, which I would like to avoid.

```python
from jax.dlpack import from_dlpack

a = cupy.arange(6).reshape((2, 3))
b = from_dlpack(a.toDlpack())
c = b[:, :2]

a.data.ptr == b.device_buffer.unsafe_buffer_pointer()  # True
a.data.ptr == c.device_buffer.unsafe_buffer_pointer()  # False
```

**Comment by zhangqiaorjc** [comment]
I think the original bug is [fixed](https://github.com/tensorflow/tensorflow/commit/28a2d0a1a83b3dfa3c903ce645aa9d319a4fd5ef).

Please open a new for avoiding the copy as an optimization if needed.
